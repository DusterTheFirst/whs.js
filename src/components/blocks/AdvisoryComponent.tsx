/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { StyleSheet, Text, View } from "react-native";

const styles = StyleSheet.create({
    container: {
        backgroundColor: "#FFFFFF",
        // borderBottomColor: "#CECECE",
        // borderBottomWidth: 1,
        height: 75,
        padding: 10,
        width: "100%"
    },
    dim: {
        color: "#808080"
    },
    dualView: {
        alignItems: "baseline",
        flexDirection: "row",
        flexWrap: "nowrap",
        justifyContent: "space-between"
    },
    info: {
        marginTop: 5
    },
    room: {
        overflow: "visible"
    },
    teacher: {
        maxWidth: "50%"
    },
    title: {
        color: "#A0A0A0",
        fontSize: 25,
        fontWeight: "bold"
    }
});

export default function AdvisoryComponent({teacher, room}: IAdvisory) {
    return (
        <View style={styles.container}>
            <View style={styles.dualView}>
                <Text style={styles.title}>Advisory</Text>
                <Text style={styles.dim}>9:38 - 9:46 AM</Text>
            </View>
            <View style={[styles.dualView, styles.info]}>
                <Text style={[styles.dim, styles.teacher]} numberOfLines={1}>{teacher}</Text>
                <Text style={[styles.dim, styles.room]} numberOfLines={1}>Room {room}</Text>
            </View>
        </View>
    );
}