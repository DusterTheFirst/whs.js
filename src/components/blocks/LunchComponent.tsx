/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import React, { memo } from "react";
import { Text, View } from "react-native";
import { classComponentStyles } from "../../styles/layout/default";
import { ITimes } from "../../util/class/extentions";

/** A component to display a lunch block */
function LunchComponent({ start, end }: ITimes) {
    return (
        <View style={classComponentStyles.container}>
            <View style={classComponentStyles.multiView}>
                <Text style={[classComponentStyles.title]}>Lunch</Text>
                <Text style={classComponentStyles.dim}>{start.format("h:mm")} - {end.format("h:mm A")}</Text>
            </View>
        </View>
    );
}

export default memo(LunchComponent);