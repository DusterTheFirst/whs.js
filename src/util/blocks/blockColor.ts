/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

/** Colors of the blocks */
export enum BlockColor {
    Red = "Red",
    Orange = "Orange",
    Yellow = "Yellow",
    Green = "Green",
    Blue = "Blue",
    Purple = "Purple",
    Tan = "Tan",
    None = "No Color"
}

/** A map of the block colors to their display colors */
type BlockColorColorMap = {
    readonly [B in BlockColor]: string;
};

/**
 * The display colors that corispond to each of the block colors
 */
const BlockColorDisplayColors: BlockColorColorMap = {
    [BlockColor.Blue]: "#2b5cbf",
    [BlockColor.Green]: "#5da539",
    [BlockColor.None]: "#999999",
    [BlockColor.Orange]: "#eb8a29",
    [BlockColor.Purple]: "#8848ad",
    [BlockColor.Red]: "#c93a3a",
    [BlockColor.Tan]: "#ad855e",
    [BlockColor.Yellow]: "#ddc923"
};

/** Lookup the display color for the given block color */
export function getDisplayColorForBlock(block: BlockColor | undefined) {
    return block === undefined ? undefined : BlockColorDisplayColors[block];
}