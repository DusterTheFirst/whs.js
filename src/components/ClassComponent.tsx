/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { Moment } from "moment";
import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { BlockColor, BlockColorDisplayColors } from "../util/blocks/blockColor";

const styles = StyleSheet.create({
    container: {
        backgroundColor: "#FFFFFF",
        // borderBottomColor: "#CECECE",
        // borderBottomWidth: 1,
        height: 75,
        padding: 10,
        width: "100%"
    },
    dim: {
        color: "#808080"
    },
    dualView: {
        alignItems: "baseline",
        flexDirection: "row",
        flexWrap: "nowrap",
        justifyContent: "space-between"
    },
    info: {
        marginTop: 5
    },
    room: {
        overflow: "visible"
    },
    teacher: {
        maxWidth: "50%"
    },
    title: {
        color: "#A0A0A0",
        fontSize: 25,
        fontWeight: "bold"
    }
});

export interface IClass {
    teacher: string;
    room: number;
    name: string;
    block: BlockColor;
    start: Moment;
    end: Moment;
}
export default function ClassComponent({teacher, room, name, start, end, block}: IClass) {
    return (
        <View style={styles.container}>
            <View style={styles.dualView}>
                <Text style={[styles.title, {color: BlockColorDisplayColors[BlockColor[block]]}]}>{name}</Text>
                <Text style={styles.dim}>{start.format("hh:mm")} - {end.format("hh:mm")}</Text>
            </View>
            <View style={[styles.dualView, styles.info]}>
                <Text style={[styles.dim, styles.teacher]} numberOfLines={1}>{teacher}</Text>
                <Text style={[styles.dim, styles.room]} numberOfLines={1}>Room {room}</Text>
            </View>
        </View>
    );
}