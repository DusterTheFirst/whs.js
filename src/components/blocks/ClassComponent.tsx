/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import React, { memo } from "react";
import { Text, View } from "react-native";
import { classComponentStyles } from "../../layout/default";
import { LabelPosition } from "../../stores/preferencesStore";
import { getDisplayColorForBlock } from "../../util/blocks/blockColor";
import { ITimes } from "../../util/class/extentions";
import { IAdvisedClass, IColored, INamed } from "../../util/class/primitives";
import usePreferences from "../../util/hooks/usePreferences";

/** A component that will display a class */
function ClassComponent({ end, start, name, block, teacher, room }: ITimes & IAdvisedClass & IColored & INamed) {
    const preferences = usePreferences();

    return (
        <View style={classComponentStyles.container}>
            <View style={classComponentStyles.multiView}>
                <Text style={[classComponentStyles.title, { color: getDisplayColorForBlock(block) }]}>{name.length === 0 ? "No Name" : name}</Text>
                <Text style={classComponentStyles.dim}>{start.format("h:mm")} - {end.format("h:mm A")}</Text>
            </View>
            <View style={[classComponentStyles.multiView, classComponentStyles.info]}>
                <View style={[{ flexDirection: "row" }, classComponentStyles.multiLeft]}>
                    <Text style={[classComponentStyles.dim, classComponentStyles.multiLeft]} numberOfLines={1}>{teacher.length === 0 ? "No Teacher" : teacher}</Text>
                    {preferences.accessibility.labelColors && preferences.accessibility.labelPosition === LabelPosition.Left ? <Text style={[classComponentStyles.dim, classComponentStyles.multiLeft, classComponentStyles.colorblindColor, preferences.accessibility.matchLabelColors ? { color: getDisplayColorForBlock(block) } : undefined]}>{block}</Text> : null}
                </View>
                {preferences.accessibility.labelColors && preferences.accessibility.labelPosition === LabelPosition.Center ? <Text style={[classComponentStyles.dim, classComponentStyles.multiMiddle, classComponentStyles.colorblindColor, preferences.accessibility.matchLabelColors ? { color: getDisplayColorForBlock(block) } : undefined]}>{block}</Text> : null}
                <View style={[{ flexDirection: "row" }, classComponentStyles.multiRight]}>
                    {preferences.accessibility.labelColors && preferences.accessibility.labelPosition === LabelPosition.Right ? <Text style={[classComponentStyles.dim, classComponentStyles.multiRight, classComponentStyles.colorblindColor, preferences.accessibility.matchLabelColors ? { color: getDisplayColorForBlock(block) } : undefined]}>{block}</Text> : null}
                    <Text style={[classComponentStyles.dim, classComponentStyles.multiRight]} numberOfLines={1}>{room.length === 0 ? "No Room" : `${isNaN(parseInt(room, 10)) ? "" : "Room "}${room}`}</Text>
                </View>
            </View>
        </View>
    );
}

export default memo(ClassComponent);