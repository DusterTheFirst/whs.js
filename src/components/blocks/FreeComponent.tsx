/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import React, { memo } from "react";
import { Text, View } from "react-native";
import { classComponentStyles } from "../../layout/default";
import { LabelPosition } from "../../stores/preferencesStore";
import { getDisplayColorForBlock } from "../../util/blocks/blockColor";
import { ITimes } from "../../util/class/extentions";
import { IColored } from "../../util/class/primitives";
import usePreferences from "../../util/hooks/usePreferences";

/** A component to display a free block */
function FreeComponent({ start, end, block }: ITimes & IColored) {
    const preferences = usePreferences();

    const posStyle = preferences.accessibility.labelPosition === LabelPosition.Left ?
        classComponentStyles.multiLeft : preferences.accessibility.labelPosition === LabelPosition.Center ?
        classComponentStyles.multiMiddle : classComponentStyles.multiRight;

    return (
        <View style={classComponentStyles.container}>
            <View style={classComponentStyles.multiView}>
                <Text style={[classComponentStyles.title, { color: getDisplayColorForBlock(block) }]}>Free</Text>
                <Text style={classComponentStyles.dim}>{start.format("h:mm")} - {end.format("h:mm A")}</Text>
            </View>
            <View style={[classComponentStyles.multiView, classComponentStyles.info]}>
                {preferences.accessibility.labelColors ? <Text style={[classComponentStyles.dim, posStyle, classComponentStyles.colorblindColor, preferences.accessibility.matchLabelColors ? { color: getDisplayColorForBlock(block) } : undefined]}>{block}</Text> : null}
            </View>
        </View>
    );
}

export default memo(FreeComponent);