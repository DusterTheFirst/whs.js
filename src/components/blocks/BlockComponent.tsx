/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React, { memo } from "react";
import { IDR, IMajor, IMinor } from "../../util/class/classes";
import { ITimes } from "../../util/class/extentions";
import { IColored } from "../../util/class/primitives";
import { isMajor, isMinor } from "../../util/class/type";
import ClassComponent from "./ClassComponent";
import FreeComponent from "./FreeComponent";

/** The parameters to pass to a BlockComponent */
interface IDisplayBlock extends IColored, ITimes {
    /** The class to display */
    clazz?: IMajor | IMinor | IDR;
}

/** A component that will display a block of many variety */
function BlockComponent({ end, start, clazz, block }: IDisplayBlock) {
    if (clazz === undefined) {
        return <FreeComponent start={start} end={end} block={block} />;
    } else {
        if (isMajor(clazz) || isMinor(clazz)) {
            return (
                <ClassComponent
                    start={start}
                    end={end}
                    name={clazz.name}
                    block={clazz.block}
                    room={clazz.room}
                    teacher={clazz.teacher}
                />
            );
        } else {
            return (
                <ClassComponent
                    start={start}
                    end={end}
                    name="Directed Reaserch"
                    block={clazz.block}
                    room={clazz.room}
                    teacher={clazz.teacher}
                />
            );
        }
    }
}

export default memo(BlockComponent);