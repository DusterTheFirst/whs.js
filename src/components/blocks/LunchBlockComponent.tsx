/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import { IClass } from "@whsha/classes/v2/class";
import { Lunch } from "@whsha/classes/v2/lunch";
import dayjs from "dayjs";
import { default as useCustomFormat } from "dayjs/plugin/customParseFormat";
import React, { memo } from "react";
import BlockComponent from "./BlockComponent";
import LunchComponent from "./LunchComponent";

dayjs.extend(useCustomFormat);

/** The parameters to pass to a LunchBlockComponent */
interface ILunchBlock {
    /** The class to display */
    lunch: Lunch;
    /** The class to display */
    clazz?: IClass;
    /** The block color */
    block: BlockColor;
    /** The timings for the classes */
    // times: {
    //     /** The first part of the lunch block */
    //     A: ITimes;
    //     /** The second part of the lunch block */
    //     B: ITimes;
    //     /** The third part of the lunch block */
    //     C: ITimes;
    // };
}

// FIXME: Correct timings
/** The timings for the lunch sections */
const LUNCH_TIMES = {
    A: {
        end: dayjs("11:22 AM", "h:mm A"),
        start: dayjs("10:55 AM", "h:mm A")
    },
    B: {
        end: dayjs("11:52 AM", "h:mm A"),
        start: dayjs("11:25 AM", "h:mm A")
    },
    C: {
        end: dayjs("12:22 PM", "h:mm A"),
        start: dayjs("11:55 AM", "h:mm A")
    }
};

/** A component that will display a block of many variety */
function LunchBlockComponent({ lunch, clazz, block }: ILunchBlock) {
    if (lunch === Lunch.First) {
        return (
            <>
                <LunchComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.A.end} />
                <BlockComponent start={LUNCH_TIMES.B.start} end={LUNCH_TIMES.C.end} block={block} clazz={clazz} />
            </>
        );
    } else if (lunch === Lunch.Second) {
        return (
            <>
                <BlockComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.A.end} block={block} clazz={clazz} />
                <LunchComponent start={LUNCH_TIMES.B.start} end={LUNCH_TIMES.B.end} />
                <BlockComponent start={LUNCH_TIMES.C.start} end={LUNCH_TIMES.C.end} block={block} clazz={clazz} />
            </>
        );
    } else if (lunch === Lunch.Third) {
        return (
            <>
                <BlockComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.B.end} block={block} clazz={clazz} />
                <LunchComponent start={LUNCH_TIMES.C.start} end={LUNCH_TIMES.C.end} />
            </>
        );
    } else {
        return (
            <BlockComponent start={LUNCH_TIMES.A.start} end={LUNCH_TIMES.C.end} block={block} clazz={clazz} />
        );
    }
}

export default memo(LunchBlockComponent);