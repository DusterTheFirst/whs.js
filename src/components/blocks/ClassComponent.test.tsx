/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import dayjs from "dayjs";
import useCustomFormat from "dayjs/plugin/customParseFormat";
import React from "react";
import { render } from "react-native-testing-library";
import { BlockColor } from "../../util/blocks/blockColor";
import ClassComponent from "./ClassComponent";

dayjs.extend(useCustomFormat);

describe("Tests different variations of <ClassComponent>", () => {
    it("Renders ClassComponent with a string room", () => {
        let comp = render(<ClassComponent name="Class" room="Office" teacher="Mr. Teacher Man" block={BlockColor.Red} start={dayjs("7:30 AM", "h:mm A")} end={dayjs("8:30 AM", "h:mm A")}/>);

        expect(comp.toJSON()).toMatchSnapshot();
    });
    it("Renders ClassComponent with a numerical room", () => {
        let comp = render(<ClassComponent name="Class" room="Office" teacher="Mr. Teacher Man" block={BlockColor.Red} start={dayjs("12:14 PM", "h:mm A")} end={dayjs("1:24 PM", "h:mm A")}/>);

        expect(comp.toJSON()).toMatchSnapshot();
    });
});