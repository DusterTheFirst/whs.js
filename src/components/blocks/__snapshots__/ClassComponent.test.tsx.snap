// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests different variations of <ClassComponent> Renders ClassComponent with a numerical room 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "height": 75,
      "justifyContent": "center",
      "padding": 10,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "stretch",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#A0A0A0",
            "fontSize": 25,
            "fontWeight": "bold",
          },
          Object {
            "color": "red",
          },
        ]
      }
    >
      Class
    </Text>
    <Text
      style={
        Object {
          "color": "#808080",
        }
      }
    >
      12:14
       - 
      1:24 PM
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignContent": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
        Object {
          "marginTop": 5,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "flex": 1,
            "textAlign": "left",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
            },
            Object {
              "flex": 1,
              "textAlign": "left",
            },
          ]
        }
      >
        Mr. Teacher Man
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "flex": 1,
            "textAlign": "right",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
            },
            Object {
              "flex": 1,
              "textAlign": "right",
            },
          ]
        }
      >
        Office
      </Text>
    </View>
  </View>
</View>
`;

exports[`Tests different variations of <ClassComponent> Renders ClassComponent with a string room 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "height": 75,
      "justifyContent": "center",
      "padding": 10,
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignContent": "stretch",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "space-between",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#A0A0A0",
            "fontSize": 25,
            "fontWeight": "bold",
          },
          Object {
            "color": "red",
          },
        ]
      }
    >
      Class
    </Text>
    <Text
      style={
        Object {
          "color": "#808080",
        }
      }
    >
      7:30
       - 
      8:30 AM
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "alignContent": "stretch",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
        Object {
          "marginTop": 5,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "flex": 1,
            "textAlign": "left",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
            },
            Object {
              "flex": 1,
              "textAlign": "left",
            },
          ]
        }
      >
        Mr. Teacher Man
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "flex": 1,
            "textAlign": "right",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
            },
            Object {
              "flex": 1,
              "textAlign": "right",
            },
          ]
        }
      >
        Office
      </Text>
    </View>
  </View>
</View>
`;
